File: ios/Gogai/Network/Endpoint.swift
import Foundation

struct Endpoint: Sendable {
    let path: String
    let method: HTTPMethod
    let queryItems: [URLQueryItem]?
    let body: Data?

    enum HTTPMethod: String, Sendable {
        case get = "GET"
        case post = "POST"
        case put = "PUT"
        case delete = "DELETE"
    }

    init(
        path: String,
        method: HTTPMethod = .get,
        queryItems: [URLQueryItem]? = nil,
        body: Data? = nil
    ) {
        self.path = path
        self.method = method
        self.queryItems = queryItems
        self.body = body
    }

    static func get(_ path: String, queryItems: [URLQueryItem]? = nil) -> Endpoint {
        Endpoint(path: path, method: .get, queryItems: queryItems)
    }

    static func post(_ path: String) -> Endpoint {
        Endpoint(path: path, method: .post)
    }

    static func post<B: Encodable & Sendable>(_ path: String, body: B) throws -> Endpoint {
        let data = try JSONEncoder().encode(body)
        return Endpoint(path: path, method: .post, body: data)
    }

    static func put<B: Encodable & Sendable>(_ path: String, body: B) throws -> Endpoint {
        let data = try JSONEncoder().encode(body)
        return Endpoint(path: path, method: .put, body: data)
    }

    static func delete(_ path: String) -> Endpoint {
        Endpoint(path: path, method: .delete)
    }

    func urlRequest(baseURL: URL) throws -> URLRequest {
        guard var components = URLComponents(
            url: baseURL.appendingPathComponent(path),
            resolvingAgainstBaseURL: false
        ) else {
            throw APIError.invalidURL
        }
        components.queryItems = queryItems?.isEmpty == false ? queryItems : nil
        guard let url = components.url else {
            throw APIError.invalidURL
        }
        var request = URLRequest(url: url)
        request.httpMethod = method.rawValue
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.httpBody = body
        return request
    }
}

File: android/app/src/main/kotlin/com/mtkg/gogai/network/Endpoint.kt
package com.mtkg.gogai.network

/// API リクエスト定義（iOS Endpoint の移植）。body は JSON 文字列。
data class Endpoint(
    val path: String,
    val method: HttpMethod = HttpMethod.GET,
    val queryItems: List<Pair<String, String>> = emptyList(),
    val body: String? = null,
    val headers: Map<String, String> = emptyMap(),
) {
    enum class HttpMethod { GET, POST, PUT, PATCH, DELETE }

    companion object {
        fun get(path: String, queryItems: List<Pair<String, String>> = emptyList(), headers: Map<String, String> = emptyMap()) =
            Endpoint(path, HttpMethod.GET, queryItems, headers = headers)

        fun post(path: String, body: String? = null, headers: Map<String, String> = emptyMap()) =
            Endpoint(path, HttpMethod.POST, body = body, headers = headers)

        fun put(path: String, body: String) = Endpoint(path, HttpMethod.PUT, body = body)

        fun patch(path: String, body: String) = Endpoint(path, HttpMethod.PATCH, body = body)

        fun delete(path: String) = Endpoint(path, HttpMethod.DELETE)
    }
}
